### PR TITLE
feed downstream by relay log at startup if need

### PR DESCRIPTION
**Pull request by july2993**
### What problem does this PR solve? 
support to recover at startup by relay log. 
follows #847 #849 
### What is changed and how it works?
- add a status in checkpoint
    - we can learn from the status that there are dirty data written into downstream behind commit ts or not.
- group the config items about relay
    - remove the `relay-read-buf-size`
    - rename `relay-log-size` to `max-file-size`
- make relay.Reader return the origin format  directly
    - this will be more flexible by the user of the reader
- feed downstream by relay log at startup if need.
    - mainly implement in `drainer/relay.go`
    - only fails to init the pd client will try this or just assume the upstream cluster is normal and works like before.

   

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)
- kill -9 drainer & pd while insert data into upstream
- check status is Running
- startup drainer
- check recover and status updated to be Normal

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

**Comment by suzaku** [comment]
@july2993 Please resolve conflicts.

**Comment by july2993** [comment]
> @july2993 Please resolve conflicts.

resolved

File: drainer/relay.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+			loaderClosed = true
+		}
+
+		// break once we drainer the success items return by loader.
+		if loaderClosed && successTxnC == nil {
+			break
+		}
+
+		select {
+		case sbinlog, ok := <-readerTxnsC:
```
Comment (suzaku) [inline drainer/relay.go]: Is it possible for two `sbinlog` to be received successively in this loop?

File: drainer/relay_test.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/pingcap/tidb-binlog/drainer/translator"
+	"github.com/pingcap/tidb-binlog/pkg/binlogfile"
+	"github.com/pingcap/tidb-binlog/pkg/loader"
+)
+
+type relaySuite struct{}
+
+var _ = check.Suite(&relaySuite{})
+
+type noOpLoader struct {
```
Comment (suzaku) [inline drainer/relay_test.go]: Embed the interface so that we don't have to repeat methods we don't need in this test?

File: drainer/checkpoint/checkpoint.go [hunk 3]
```diff
@@ -19,6 +19,14 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	// StatusNormal means server quit normally, data <= ts is synced to downstream
+	StatusNormal int = 0
```
Comment (GregoryIan) [inline drainer/checkpoint/checkpoint.go]: Worth a better name, like `StatusConsistent` and `StatusInconsistent`

Comment (GregoryIan) [reply drainer/checkpoint/checkpoint.go]: Good Change!

File: drainer/checkpoint/mysql.go [hunk 4]
```diff
@@ -68,6 +74,16 @@ func newMysql(cfg *Config) (CheckPoint, error) {
 		return nil, errors.Annotatef(err, "exec failed, sql: %s", sql)
 	}
 
+	if sp.clusterID == 0 {
+		id, err := getClusterID(db, sp.schema, sp.table)
```
Comment (GregoryIan) [inline drainer/checkpoint/mysql.go]: This logic is very weird. 
In order to handle the situation where the upstream cluster is completely suspended, drainer needs a correct way to fetch the  cluster id. But current implementation is not satisfying, like drainer can't allow store multiple entries in checkpoint table

Comment (GregoryIan) [reply drainer/checkpoint/mysql.go]: I file an issue about it https://github.com/pingcap/tidb-binlog/issues/889

File: drainer/checkpoint/util.go [hunk 5]
```diff
@@ -14,12 +14,18 @@
...
 	"fmt"
 
 	// mysql driver
 	_ "github.com/go-sql-driver/mysql"
+	"github.com/pingcap/errors"
 )
 
+// ErrNoCheckpointItem represents there's no any checkpoint item and the cluster id must be specified
+// for the mysql checkpoint type.
+var ErrNoCheckpointItem = errors.New("no any checkpoint item")
```
[the remark this answers is not in the document]
Comment (GregoryIan) [reply drainer/checkpoint/util.go]: Is there any special reason @suzaku ?  Looking forward to learning something from your comment 😄

File: drainer/checkpoint/util_test.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	tests := []struct {
+		name              string
+		rows              []uint64
+		id                uint64
+		err               bool
+		checkSpecifiedErr error
+	}{
+		{"no row", nil, 0, true, ErrNoCheckpointItem},
+		{"on row", []uint64{1}, 1, false, nil},
+		{"multi row", []uint64{1, 2}, 0, true, nil},
```
Comment (GregoryIan) [inline drainer/checkpoint/util_test.go]: should the `checkSpecifiedErr` be nil?

File: drainer/config.go [hunk 7]
```diff
@@ -139,9 +148,6 @@ func NewConfig() *Config {
 	fs.StringVar(&cfg.SyncerCfg.IgnoreSchemas, "ignore-schemas", "INFORMATION_SCHEMA,PERFORMANCE_SCHEMA,mysql", "disable sync those schemas")
 	fs.IntVar(&cfg.SyncerCfg.WorkerCount, "c", 16, "parallel worker count")
 	fs.StringVar(&cfg.SyncerCfg.DestDBType, "dest-db-type", "mysql", "target db type: mysql or tidb or file or kafka; see syncer section in conf/drainer.toml")
-	fs.StringVar(&cfg.SyncerCfg.RelayLogDir, "relay-log-dir", "", "path to relay log of syncer")
```
Comment (GregoryIan) [inline drainer/config.go]: can we leave them? 

they may improve ease of use, like in K8s, because configuration files are always cumbersome

File: pkg/util/util.go [hunk 8]
```diff
@@ -251,20 +252,12 @@ func GetPdClient(etcdURLs string, securityConfig security.Config) (pd.Client, er
 		return nil, errors.Trace(err)
 	}
 
-	pdReconnTimes := 30
-
 	var pdCli pd.Client
-	for i := 1; i < pdReconnTimes; i++ {
```
Comment (GregoryIan) [inline pkg/util/util.go]: Removing retries may reduce usability, as long as drainer can't connect to pd, it will try the relay log and then exit.

Comment (GregoryIan) [reply pkg/util/util.go]: we can leave the simple retry logic

File: drainer/checkpoint/util.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (suzaku) [reply drainer/checkpoint/util.go]: @GregoryIan When returning an error created using the `pingcap/errors` package, the stack trace would end in the line where the error is declared.

File: drainer/checkpoint/util_test.go (hunk 6 above)
Comment (july2993) [reply drainer/checkpoint/util_test.go]: yes,  true, nil means the result will be some error but we don't check what kind of error.

File: pkg/util/util.go (hunk 8 above)
Comment (july2993) [reply pkg/util/util.go]: As the commen added, the pd.NewClient will retry 100 time already.

File: drainer/relay.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	if cp.Status() == checkpoint.StatusNormal {
+		return nil
+	}
+
+	reader, err := relay.NewReader(scfg.Relay.LogDir, 1 /* readBufferSize */)
+	if err != nil {
+		return errors.Annotate(err, "failed to create reader")
+	}
+
+	db, ld, err := sync.CreateLoader(scfg.To, scfg.WorkerCount, scfg.TxnBatch,
```
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/relay.go]: ? should check err first, then defer

File: drainer/config.go [hunk 10]
```diff
@@ -74,14 +74,23 @@ type SyncerConfig struct {
...
 }
 
+// RelayConfig is the Relay log's configuration.
+type RelayConfig struct {
+	LogDir      string `toml:"log-dir" json:"log-dir"`
+	MaxFileSize int64  `toml:"max-file-size" json:"max-file-size"`
+}
+
+// SwitchOn return true if we need to handle relay log.
+func (rc RelayConfig) SwitchOn() bool {
```
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/config.go]: changed

File: drainer/config.go (hunk 7 above)
Comment (july2993) [reply drainer/config.go]: reverted

File: drainer/relay.go (hunk 1 above)
Comment (july2993) [reply drainer/relay.go]: fixed

File: drainer/checkpoint/util.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/checkpoint/util.go]: use std in https://github.com/pingcap/tidb-binlog/pull/883/commits/d34142569b67bb097232bd91c791dce05c665f38

File: drainer/relay.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+			ld.Close()
+			loaderClosed = true
+		}
+
+		// break once we drainer the success items return by loader.
+		if loaderClosed && successTxnC == nil {
+			break
+		}
+
+		select {
```
Comment (suzaku) [inline drainer/relay.go]: The logic in this `select` is a bit hard to follow and reason about. Will it be simpler if we just we two separate goroutine for reading and writing?

File: drainer/relay_test.go (hunk 2 above)
Comment (july2993) [reply drainer/relay_test.go]: all need unless `SaveMode`, I will keep this.
instead, we should make the pkg/Loader easy to test(not make it an interface only for test), we can use the struct directly and add some method like set it silently accept `input` and return success directly.

File: drainer/sync/mysql.go [hunk 12]
```diff
@@ -41,11 +41,19 @@ type MysqlSyncer struct {
...
+// CreateLoader create the Loader instance.
+func CreateLoader(
+	cfg *DBConfig,
+	worker int,
+	batchSize int,
+	queryHistogramVec *prometheus.HistogramVec,
+	sqlMode *string,
+	destDBType string,
+	info *loopbacksync.LoopBackSync,
+) (db *sql.DB, ld loader.Loader, err error) {
```
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/sync/mysql.go]: update in https://github.com/pingcap/tidb-binlog/pull/883/commits/bb81853698b4d86fb8d81481b49b48d31e674c68

File: drainer/relay.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	checkpointTS := cp.TS()
+	lastSuccessTS := checkpointTS
+	r.Run()
+
+	loaderQuit := make(chan struct{})
+	var loaderErr error
+	go func() {
+		ld.SetSafeMode(true)
+		loaderErr = ld.Run()
+		close(loaderQuit)
```
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/relay.go]:
```
120         case success, ok := <-successTxnC:
121             if !ok {
122                 successTxnC = nil
123                 log.Info("success closed")
124                 continue
125             }
126             lastSuccessTS = success.Metadata.(int64)
``` 
will continue the loop and run into the `loaderQuit` case at nest select

File: drainer/relay.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+		case <-loaderQuit:
+			if loaderErr != nil {
+				return errors.Trace(loaderErr)
+			}
+		}
+	}
+
+	log.Info("finish feed by relay log")
+
+	readerErr := <-r.Error()
```
[the remark this answers is not in the document]
Comment (july2993) [reply drainer/relay.go]: `feedByRelayLog` will need to close `ld` once having put all txns into it, and  `r.Error()` should better put here together with the according `r.Run()` and `r.Txns()`

File: drainer/checkpoint/checkpoint.go (hunk 3 above)
Comment (july2993) [reply drainer/checkpoint/checkpoint.go]: rename as `StatusConsistent` https://github.com/pingcap/tidb-binlog/pull/883/commits/002d4493a4f5cfcb06bd41acc762788271b5c2d9

File: drainer/relay.go (hunk 11 above)
Comment (july2993) [reply drainer/relay.go]: actually first try in multi goroutine style first, but finally, give up, will keep the code style now.

File: pkg/util/util.go (hunk 8 above)
Comment (GregoryIan) [reply pkg/util/util.go]: now it's reasonable
